### PR TITLE
8292327: java.io.EOFException in InflaterInputStream after JDK-8281962

### PR DESCRIPTION
**Pull request by simonis**
The problem is that after [JDK-8281962](https://bugs.openjdk.org/browse/JDK-8281962) we call `fill()` unconditionally (and before calling `Inflater::inflate()`) in `InflaterInputStream::read()` if `Inflater::needsInput()` is true. This misses the case where the native inflater has consumed all its input (i.e. `Inflater::needsInput()` returns true) but hasn't managed to write all the inflated data into the output buffer (e.g. because it was to small). In rare cases, there might be now more input available (i.e. calling `InflaterInputStream::fill()` will throw an `EOFException`) but we still have to call `Inflater::inflate()` to consume the buffered output from the underlying native inflater until inflation stops. 

The documentation of the `inflate()` method in `zlib.h` mentions this explicitely:

> "If `inflate()` returns `Z_OK` and with zero `avail_out`, it must be called again after making room in the output buffer because there might be more output pending."

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292327](https://bugs.openjdk.org/browse/JDK-8292327): java.io.EOFException in InflaterInputStream after JDK-8281962

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9881/head:pull/9881` \
`$ git checkout pull/9881`

Update a local copy of the PR: \
`$ git checkout pull/9881` \
`$ git pull https://git.openjdk.org/jdk pull/9881/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9881`

View PR using the GUI difftool: \
`$ git pr show -t 9881`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9881.diff">https://git.openjdk.org/jdk/pull/9881.diff</a>

</details>

### CODE REVIEW

File: test/jdk/java/util/zip/InflaterInputStream/EOF.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    // a 128 byte read followed by a 512 byte read fails after JDK-8281962
+    InflaterInputStream is =
+        new InflaterInputStream(new ByteArrayInputStream(bytes), new Inflater(/*nowrap=*/ true));
+    byte[] buf = new byte[512];
+    int n = is.read(buf, 0, 128);
+    while (n > 0) {
+      n = is.read(buf, 0, 1); // Unexpected end of ZLIB input stream
+    }
+  }
```
Comment (AlanBateman) [inline test/jdk/java/util/zip/InflaterInputStream/EOF.java:61]: Can you re-format this to use 4 space indent?

File: src/java.base/share/classes/java/util/zip/InflaterInputStream.java [hunk 2]
```diff
@@ -155,7 +155,12 @@ public int read(byte[] b, int off, int len) throws IOException {
...
                     return -1;
                 }
-                if (inf.needsInput()) {
+                if (inf.needsInput() && !inf.hasPendingOutput()) {
+                    // Even if needsInput() is true, the native inflater may have some
+                    // buffered data which couldn't fit in to the output buffer during the
+                    // last call to inflate. Consume that bffered data first before calling
+                    // fill() to avoid an EOF error if no more input is available and the
+                    // next call to inflate will finish the inflation.
                     fill();
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/InflaterInputStream.java:164]: We need to submit a bug against fill() too as it is not currently specified to throw when the input is at EOF.

File: src/java.base/share/classes/java/util/zip/InflaterInputStream.java [hunk 3]
```diff
@@ -155,7 +155,12 @@ public int read(byte[] b, int off, int len) throws IOException {
                     reachEOF = true;
                     return -1;
                 }
-                if (inf.needsInput()) {
+                if (inf.needsInput() && !inf.hasPendingOutput()) {
+                    // Even if needsInput() is true, the native inflater may have some
+                    // buffered data which couldn't fit in to the output buffer during the
+                    // last call to inflate. Consume that bffered data first before calling
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/InflaterInputStream.java:161]: bffered -> buffered

File: test/jdk/java/util/zip/InflaterInputStream/EOF.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ *          still has unwritten output in its internal buffer.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.zip.Inflater;
+import java.util.zip.InflaterInputStream;
+
+public class EOF {
```
Comment (AlanBateman) [inline test/jdk/java/util/zip/InflaterInputStream/EOF.java:37]: This is not a general purpose test for EOF so I think we'll need to find a better name.  BufferedBytesAtInputEOF might work.

Also the test doesn't check the output. We'll need to do that to make it a more complete test.

File: src/java.base/share/classes/java/util/zip/InflaterInputStream.java [hunk 5]
```diff
@@ -155,7 +155,12 @@ public int read(byte[] b, int off, int len) throws IOException {
                     reachEOF = true;
                     return -1;
                 }
-                if (inf.needsInput()) {
+                if (inf.needsInput() && !inf.hasPendingOutput()) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/InflaterInputStream.java:158]: I think the approach looks okay and avoids attempting to fill and the unspecified EOFException that it throws when the input is at EOF.

File: test/jdk/java/util/zip/InflaterInputStream/EOF.java (hunk 1 above)
Comment (simonis) [reply test/jdk/java/util/zip/InflaterInputStream/EOF.java:61]: Done

File: src/java.base/share/classes/java/util/zip/InflaterInputStream.java (hunk 2 above)
Comment (simonis) [reply src/java.base/share/classes/java/util/zip/InflaterInputStream.java:164]: Not sure about this one? `fill()` is specified to throw an `IOException` and `EOFException` is an `IOException`. It probably depends on how you interpret the current "*if an I/O error has occurred*" description in the throws section.

Do you propose to be more specific there and write something like "*if an I/O error has occurred or the end of the input buffer has been reached*"?

File: src/java.base/share/classes/java/util/zip/InflaterInputStream.java (hunk 3 above)
Comment (simonis) [reply src/java.base/share/classes/java/util/zip/InflaterInputStream.java:161]: Fixed.

File: test/jdk/java/util/zip/InflaterInputStream/EOF.java (hunk 4 above)
Comment (simonis) [reply test/jdk/java/util/zip/InflaterInputStream/EOF.java:37]: Changed the name to `UnexpectedEndOfZlibStream.java` and updated the test as requested.

**Comment by AlanBateman** [comment]
> The problem is that after [JDK-8281962](https://bugs.openjdk.org/browse/JDK-8292327) 

In the description, this should link to JDK-8281962 rather than (JDK-8292327). Just pointing out as someone looking at this PR in the future might get confused as to which issue caused the regression.

File: src/java.base/share/classes/java/util/zip/InflaterInputStream.java (hunk 2 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/util/zip/InflaterInputStream.java:164]: > Not sure about this one? `fill()` is specified to throw an `IOException` and `EOFException` is an `IOException`. It probably depends on how you interpret the current "_if an I/O error has occurred_" description in the throws section.
> 
> Do you propose to be more specific there and write something like "_if an I/O error has occurred or the end of the input buffer has been reached_"?

Like many JDK 1.1 era APIs, InflaterInputStream is very under-specified. In this case, the protected fill method is begging us to specify how it behaves when the input is at EOF. An I/O error is surprising here as it sets the protected len field to indicate EOF. We can't change it after 25 years so we will need to specify existing behavior.

Comment (simonis) [reply src/java.base/share/classes/java/util/zip/InflaterInputStream.java:164]: OK, I've opened [JDK-8292427: Improve specification of InflaterInputStream.fill()](https://bugs.openjdk.org/browse/JDK-8292427) to track this.

**Comment by simonis** [comment]
I've updated the link in the description (sorry, copy-paste error) and opened a JBS issue for `fill()`.

What else is needed to push this to HEAD?

And what do we want to do about JDK 19? Have you read my [last comment](https://bugs.openjdk.org/browse/JDK-8292327?focusedCommentId=14517812&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14517812) in the JBS issue`?

File: test/jdk/java/util/zip/InflaterInputStream/UnexpectedEndOfZlibStream.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        byte[] bytes = {
+            85, -115, 49, 14, -125, 48, 12, 0, -9, -68, 2, 49, -63, 98, -87, 93, 89, 90, -47, -103, 14,
+            -99, 58, -102, 96, 5, -89, 36, -74, -120, -123, -44, -33, -61, 8, -21, -23, 78, -9, 120, -113,
+            -111, -68, -15, 70, -3, -128, -119, -102, -6, -11, 29, 62, 79, -27, -70, 117, -118, -2, -121,
+            -127, 42, 47, 9, -126, 72, 88, 8, 80, -75, -64, -12, -49, -104, -40, 8, -74, 27, -108, 25, 87,
+            -102, 14, -50, -99, 115, -100, 84, 86, 59, -5, -15, 46, 99, -12, -128, 57, -117, -95, -79,
+            -28, 2, -41, -33, 81, -19,
+        };
+        String deflated = "@ObjectiveCName(\"DYNSApi\")\npackage com.google.apps.dynamite.v1.shared.api;\n\n"+
+            "import com.google.j2objc.annotations.ObjectiveCName;\n\n";
```
Comment (AlanBateman) [inline test/jdk/java/util/zip/InflaterInputStream/UnexpectedEndOfZlibStream.java:48]: Can you replace the deflated text with something that doesn't look like it came from a google repository?

**Comment by AlanBateman** [comment]
> And what do we want to do about JDK 19? 

It's not a P1 so I don't think it's a change for the RC builds. It is a regression in 19 so we will need to create a RN-KnownIssue release note (I will do that).

I agree with Lance's suggestion to revert JDK-8281962 in JDK 19u. It might be that in the future that you want to re-do the original change for a future 19 update but I think we should give the changes in the main line time to bake.

Separately, we need a high priority issue to create tests for this area. We also need to fix the specification of the fill method as we've been discussing here.




- Revert change for 19.0.1 to fix regression
- Fix issue in mainline for JDK 20. Give time to bake befoer thinking about putting the change+fix into a JDK 19 update.
- Create issue to fix InflaterInputStream.fill
- Create high priority issue to add more tests for Inflate + friends

File: test/jdk/java/util/zip/InflaterInputStream/UnexpectedEndOfZlibStream.java (hunk 6 above)
Comment (simonis) [reply test/jdk/java/util/zip/InflaterInputStream/UnexpectedEndOfZlibStream.java:48]: First of all I just realized that `deflated` is the wrong name, it should actually be `inflated`.

The deflated source (in `bytes`) is the payload which triggers the bug and I can't simply change the string to which it decompresses :)

I'll try to find another, more innocent string such that its compressed version also triggers the bug.

Comment (simonis) [reply test/jdk/java/util/zip/InflaterInputStream/UnexpectedEndOfZlibStream.java]: Last commit contains as more neutral payload text.

Comment (AlanBateman) [reply test/jdk/java/util/zip/InflaterInputStream/UnexpectedEndOfZlibStream.java]: Thanks for the update to change the payload.

**Review by AlanBateman — APPROVED** [review]

**Comment by simonis** [comment]
Thanks for the review @AlanBateman.

I'll wait with the integration until tomorrow to also give the colleagues in the US time zone a chance to look at the PR.

**Review by LanceAndersen — APPROVED** [review]
This looks OK Volker.

It would be nice if possible to add another test or two prior to pushing as an extra sanity check.  If not, a follow on would be ideal.

I assume you have run Mach1-3 and the JCK tests for extra sanity?

**Comment by simonis** [comment]
Hi @LanceAndersen,

Thanks for the review. I can't run Mach 1-3 because it's Oracle internal but I've run the corresponding JCK tests.

I think more and better tests should be done as a follow-up task as suggested by Alan.

**Pull request closed without merging**